### ISSUE TITLE
Assist: group imports

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#2220]
```rust
use std::fs; // <- cursor is on this line
use std::io::prelude::*;
use std::io::BufReader;
use std::path::Path;
```

->

```rust
use std::{
    fs,
    io::prelude::*,
    io::BufReader,
    path::Path,
};
```

A similar, but simpler assist is [split import](https://github.com/rust-analyzer/rust-analyzer/blob/f5e1b0f97c9e46b5186f99d744f4587b2aee397e/crates/ra_assists/src/assists/split_import.rs)

**Comment by max-frai** [comment]
Oh yes, we need this! :)

**Comment by matklad** [comment]
@max-frai I think it should be non-recursive for the start: it's possible to apply the assist several times manually. Longterm, what we really need here is assist groups like in intelliJ

```
💡
 |group uses      | > | recursively |
 |optimize imports|
```

**Comment by max-frai** [comment]
You are right. Also, it would be very useful to introduce: remove unused imports. Can rust-analyzer find such places?

**Comment by matklad** [comment]
not yet :(

**Comment by killercup** [comment]
For the record, rustfmt's unstable [merge_imports](https://github.com/rust-lang/rustfmt/blob/377346639b7d80b75de87acebb0652c69ce652d8/Configurations.md#merge_imports) option does exactly this.

**Comment by Matthias247** [comment]
I have played a bit around with this. Mostly to get more familiar with how rust-analyzer works. Not yet sure whether I will get it to completion :-)

One question regarding the implementation: It seems like this can be implemented in 2 ways:

**Option A: Manipulate the existing syntax tree**

We identify the common path segments of all `use` statements, and then use edit commands to insert braces and commas at the correct places and remove the `use prefix::` part of the next statement. That's what my hacky first implementation does, and where the `edit` looks like:
  ```rust
  ctx.add_assist(AssistId("group_import"), "group import", |edit| {
        edit.target(use_node.syntax().text_range());
        edit.insert(open_brace_segment.syntax().text_range().start(), "{");
        edit.insert(close_brace_segment.syntax().text_range().end(), "}");
        edit.delete(to_remove);
        edit.replace(use_semicolon_text_range, ",");
    })
  ```

The downside of this approach seeems to be that it seems to either falls down or get super complex on more complex merges. E.g. if the first statement already contains an opening brace for submodules, and we need to shift the second import into it without adding another brace. I guess also if things are already trees instead of just long import statements it gets more complicated.

**Option B: Rewrite import statements**

We parse all imports in suitable tree structure (like `hir::Path`, but I think that one only represents a single path and not an import tree), try to merge those, and if the merge and if the merge succeeds write the complete tree (which replaces all existing things).

I feel like this option would be easier to manage than trying to figure out precise necessary edits. But maybe I'm overestimating the effort. The downside of **Option B** is certainly that it's more invasive, and if the rewriting part does not support all AST constructs it would be a lossy and buggy operation.

**Comment by matklad** [comment]
Yeah, "how to edit syntax trees" is an open question. I've been doing this since 2015 and still don't know the answer :-) 

One general rule of thumb is that it's better re-use, rather than re-write bits written by the user. For example, if the user did `use std::collections::{HashMap /* FIXME: use hashbrown? */, BTreeMap}`, we ideally should not lose the comment. Approaches which try to build a fully-abstract IR usually don't handle this Besides comments, it's often a good default to preserve formatting. 

Another choice is how to do the edits. There are two approaches:
* edit the text directly, by just replacing ranges
* construct the new syntax node and *replace* an existing node with a new one

The latter usually is more verbose, but in general is easier to understand and, ideally, can handle some minor details like adding whitespace or punctuation automatically. 

I would go with the latter approach, using something like this:

```rust
fn merge(trees: Vec<ast::UseTree>) -> Option<ast::UseTree> {
  let common_prefix: ast::Path = common_prefix(&trees)?;
  let subtrees: Vec<ast::UseTree> = Vec::new();
  for tree in trees.iter() {
    let subtree = remove_prefix(&common_prefix, &tree)
    subtrees.push(subtree)
  }
  make::use_tree(common_prefix, subtrees);
}

fn common_prefix(trees: &[ast::UseTree]) -> Option<ast::Path> {}
fn remove_prefix(prefix: &ast::Path, tree: &ast::UseTree) -> ast::UseTree
```

See the `make.rs`  module for primitive building blocks of various ast elements. It doesn't have constructos for use items yet, they should be added .

**Comment by TomasKralCZ** [comment]
Is anyone still working on this ? If not, I'd like to try to implement it.

**Comment by matklad** [comment]
@TomasKralCZ not that I am aware of, so go for it!

**Comment by Matthias247** [comment]
@TomasKralCZ In case it helps you. This is the piece of code I had written to experiment with it: https://gist.github.com/Matthias247/db335901003943cac36186812a191268

It handled some cases, but it seemed in general a lot more complicated (see discussion above). I'm currently distracted by other things, so no longer working on this.

**Comment by matklad** [comment]
Closed by #3651
